File: src/main/java/alfheim/common/block/tile/TileAnomaly.java
package alfheim.common.block.tile;

import java.util.HashMap;
import java.util.List;

import alexsocol.asjlib.ASJUtilities;
import alfheim.api.block.tile.SubTileEntity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import vazkii.botania.common.block.tile.TileMod;

public class TileAnomaly extends TileMod {
	
	public static final String TAG_SUBTILE_MAIN = "subTileMain";
	public static final String TAG_SUBTILE_NAME = "subTileName";
	public static final String TAG_SUBTILE_CMP = "subTileCmp";
	public static final String TAG_SUBTILE_COUNT = "subTileCount";
	
	public HashMap<String, SubTileEntity> subTiles = new HashMap<String, SubTileEntity>();
	public String mainSubTile;
	public int compatibilityBit = 0; // not serializing because will be recalculated on load
	
	@Override
	public void updateEntity() {
		if (mainSubTile == null || mainSubTile.isEmpty() || subTiles.get(mainSubTile) == null) return;
		
		List<Object> l = subTiles.get(mainSubTile).getTargets();
		for (SubTileEntity subTile : subTiles.values()) subTile.updateEntity(l);
	}
	
	public boolean onActivated(ItemStack stack, EntityPlayer player, World world, int x, int y, int z) {
		boolean flag = false;
		for (SubTileEntity subTile : subTiles.values()) flag |= subTile.onActivated(stack, player, world, x, y, z);
		return flag;
	}
	
	public TileAnomaly addSubTile(String name) {
		return addSubTile(SubTileEntity.forName(name), name);
	}
	
	public TileAnomaly addSubTile(SubTileEntity sub, String name) {
		if (sub == null || !canAdd(sub)) return this;
		
		compatibilityBit |= sub.typeBits();
		
		if (mainSubTile == null || mainSubTile.isEmpty()) mainSubTile = name;
		
		subTiles.put(name, sub);
		return (TileAnomaly) (sub.superTile = this);
	}
	
	public boolean canAdd(SubTileEntity sub) {
		return (compatibilityBit & sub.typeBits()) == 0;
	}
	
	@Override
	public void writeCustomNBT(NBTTagCompound cmp) {
		super.writeCustomNBT(cmp);
		
		if (mainSubTile == null) return;
		
		try {
			cmp.setString(TAG_SUBTILE_MAIN, mainSubTile);
	
			int c = subTiles.keySet().size();
			cmp.setInteger(TAG_SUBTILE_COUNT, c);
			
			NBTTagCompound subCmp;
			
			for (String name : subTiles.keySet()) {
				cmp.setString(TAG_SUBTILE_NAME + c, name);
				
				subCmp = new NBTTagCompound();
				cmp.setTag(TAG_SUBTILE_CMP + (c--), subCmp);
				
				subTiles.get(name).writeToNBT(subCmp);
			}
		} catch (Throwable e) {
			ASJUtilities.error("Got exception writing anomaly data. It will be discarded.");
			e.printStackTrace();
		}
	}
	
	@Override
	public void readCustomNBT(NBTTagCompound cmp) {
		super.readCustomNBT(cmp);
		
		mainSubTile = cmp.getString(TAG_SUBTILE_MAIN);
		
		int c = cmp.getInteger(TAG_SUBTILE_COUNT);
		
		String subTileName;
		NBTTagCompound subCmp;
		SubTileEntity subTile;
		
		for (; c > 0; c--) {
			subTileName = cmp.getString(TAG_SUBTILE_NAME + c);
			subTile = SubTileEntity.forName(subTileName);
			
			subCmp = cmp.getCompoundTag(TAG_SUBTILE_CMP + c);
			if(subTile != null && !subCmp.hasNoTags())
				subTile.readFromNBT(subCmp);
			
			addSubTile(subTile, subTileName);
		}
	}
}

File: src/main/java/alfheim/common/block/tile/TileAnomaly.kt
package alfheim.common.block.tile

import alexsocol.asjlib.ASJUtilities
import alexsocol.asjlib.extendables.block.TileImmobile
import alfheim.api.block.tile.SubTileAnomalyBase
import alfheim.common.item.equipment.bauble.ItemSpatiotemporalRing
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.world.World
import java.util.*

class TileAnomaly: TileImmobile() {
	
	val subTiles = HashMap<String, SubTileAnomalyBase>()
	var mainSubTile: String? = null
	var compatibilityBit = 0 // not serializing because will be recalculated on load
	
	override fun updateEntity() {
		super.updateEntity()
		
		val main = mainSubTile
		if (main == null || main.isEmpty() || subTiles[main] == null) return
		
		val l = subTiles[main]!!.targets as MutableList<Any?>
		
		l.removeIf { it is EntityPlayer && ItemSpatiotemporalRing.hasProtection(it) }
		for (subTile in subTiles.values) subTile.updateEntity(l)
	}
	
	fun onActivated(stack: ItemStack?, player: EntityPlayer, world: World, x: Int, y: Int, z: Int): Boolean {
		var flag = false
		for (subTile in subTiles.values) flag = flag or subTile.onActivated(stack, player, world, x, y, z)
		return flag
	}
	
	fun addSubTile(name: String): TileAnomaly {
		return addSubTile(SubTileAnomalyBase.forName(name), name)
	}
	
	fun addSubTile(sub: SubTileAnomalyBase?, name: String): TileAnomaly {
		if (sub == null || !canAdd(sub)) return this
		
		compatibilityBit = compatibilityBit or sub.typeBits()
		
		if (mainSubTile == null || mainSubTile!!.isEmpty()) mainSubTile = name
		
		subTiles[name] = sub
		sub.superTile = this
		return sub.superTile as TileAnomaly
	}
	
	fun canAdd(sub: SubTileAnomalyBase): Boolean {
		return compatibilityBit and sub.typeBits() == 0
	}
	
	override fun writeCustomNBT(nbt: NBTTagCompound) {
		super.writeCustomNBT(nbt)
		
		if (mainSubTile == null) return
		
		try {
			nbt.setString(TAG_SUBTILE_MAIN, mainSubTile!!)
			
			var c = subTiles.keys.size
			nbt.setInteger(TAG_SUBTILE_COUNT, c)
			
			var subCmp: NBTTagCompound
			
			for (name in subTiles.keys) {
				nbt.setString(TAG_SUBTILE_NAME + c, name)
				
				subCmp = NBTTagCompound()
				nbt.setTag(TAG_SUBTILE_CMP + c--, subCmp)
				
				subTiles[name]!!.writeToNBT(subCmp)
			}
		} catch (e: Throwable) {
			ASJUtilities.error("Got exception writing anomaly data. It will be discarded.")
			e.printStackTrace()
		}
		
	}
	
	override fun readCustomNBT(nbt: NBTTagCompound) {
		super.readCustomNBT(nbt)
		
		mainSubTile = nbt.getString(TAG_SUBTILE_MAIN)
		
		var c = nbt.getInteger(TAG_SUBTILE_COUNT)
		
		var subTileName: String
		var subCmp: NBTTagCompound
		var subTile: SubTileAnomalyBase?
		
		while (c > 0) {
			subTileName = nbt.getString(TAG_SUBTILE_NAME + c)
			subTile = SubTileAnomalyBase.forName(subTileName)
			
			subCmp = nbt.getCompoundTag(TAG_SUBTILE_CMP + c)
			if (subTile != null && !subCmp.hasNoTags())
				subTile.readFromNBT(subCmp)
			
			addSubTile(subTile, subTileName)
			c--
		}
	}
	
	companion object {
		
		const val TAG_SUBTILE_MAIN = "subTileMain"
		const val TAG_SUBTILE_NAME = "subTileName"
		const val TAG_SUBTILE_CMP = "subTileCmp"
		const val TAG_SUBTILE_COUNT = "subTileCount"
	}
}